apps/experiments/filters.py

```diff
@@ -192,21 +192,41 @@ def build_tags_filter(operator, value):
         if not selected_tags:
             return None
         if operator == Operators.ANY_OF:
-            return Q(chat__tags__name__in=selected_tags)
+            chat_tags_condition = Q(chat__tags__name__in=selected_tags)
+            message_tags_condition = Q(chat__messages__tags__name__in=selected_tags)
+            return chat_tags_condition | message_tags_condition
+
         elif operator == Operators.ALL_OF:
-            content_type = ContentType.objects.get_for_model(Chat)
             conditions = Q()
+            chat_content_type = ContentType.objects.get_for_model(Chat)
+            chat_message_content_type = ContentType.objects.get_for_model(ChatMessage)
+
             for tag in selected_tags:
-                conditions &= Exists(
+                chat_tag_exists = Exists(
                     CustomTaggedItem.objects.filter(
                         object_id=OuterRef("chat_id"),
-                        content_type_id=content_type,
+                        content_type_id=chat_content_type.id,
                         tag__name=tag,
                     )
                 )
+                message_tag_exists = Exists(
+                    CustomTaggedItem.objects.filter(
+                        content_type_id=chat_message_content_type.id,
+                        tag__name=tag,
+                        object_id__in=Subquery(
+                            ChatMessage.objects.filter(chat_id=OuterRef(OuterRef("chat_id"))).values("id")
```

Comment:
Since `ChatMessage` has the `chat_id`, you should be able to use
```python
ChatMessage.objects.filter(chat_id=OuterRef("chat_id")).values("id")
```